### PR TITLE
Migrate legacy wallets that are not loaded

### PR DESCRIPTION
**Pull request by achow101**
Currently the Migrate Wallet menu item can only be used to migrate the currently loaded wallet. This is not suitable for the future when legacy wallets can no longer be loaded at all, but should still be able to be migrated. This PR changes that menu item into a menu list like Open Wallet and lets users migrate any legacy wallet in their wallet directory regardless of the wallets loaded.

One issue I ran into was dealing with encrypted wallets. Ideally, we would detect whether a wallet is encrypted, and prompt the user for their passphrase at that time. However, that's actually difficult to do in the GUI since migration will unload the wallet if it was already loaded, and reload it without connecting it to any signals or interfaces. Only then can it detect whether a wallet is encrypted, but then there is no `WalletModel` or even an `interfaces::Wallet` that the GUI could use to unlock it via a callback.

To deal with this, I've opted to just add a button to the migration dialog box that has the user enter their passphrase first, along with instructional text to use that button if their wallet was encrypted. If the user enters the wrong passphrase or clicked the other button that does not prompt for the passphrase, migration will fail with a message indicating that the passphrase was incorrect.

### CODE REVIEW

**Comment by hebasto** [comment]
Concept ACK.

**Comment by luke-jr** [comment]
>To deal with this, I've opted to just add a button to the migration dialog box that has the user enter their passphrase first, along with instructional text to use that button if their wallet was encrypted. 

If we're preemptively making UX changes just to avoid changing UX in the future, it seems like we shouldn't take shortcuts like this... :/

**Comment by achow101** [comment]
> > To deal with this, I've opted to just add a button to the migration dialog box that has the user enter their passphrase first, along with instructional text to use that button if their wallet was encrypted.
> 
> If we're preemptively making UX changes just to avoid changing UX in the future, it seems like we shouldn't take shortcuts like this... :/

We already do this in the RPC too, for exactly the same reason.

File: src/qt/bitcoingui.cpp [hunk 1]
```diff
@@ -456,10 +457,33 @@ void BitcoinGUI::createActions()
...
-            activity->migrate(walletFrame->currentWalletModel());
+        connect(m_migrate_wallet_menu, &QMenu::aboutToShow, [this] {
+            m_migrate_wallet_menu->clear();
+            for (const auto& [wallet_name, info] : m_wallet_controller->listWalletDir()) {
+                const auto& [loaded, format] = info;
+                QString name = GUIUtil::WalletDisplayName(wallet_name);
+                // Menu items remove single &. Single & are shown when && is in
+                // the string, but only the first occurrence. So replace only
+                // the first & with &&.
+                name.replace(name.indexOf(QChar('&')), 1, QString("&&"));
```
Comment (hebasto) [inline src/qt/bitcoingui.cpp:468]: https://github.com/bitcoin-core/gui/pull/828 can relevant to this code.

File: src/qt/askpassphrasedialog.h [hunk 2]
```diff
@@ -26,6 +26,7 @@ class AskPassphraseDialog : public QDialog
         Encrypt,    /**< Ask passphrase twice and encrypt */
         Unlock,     /**< Ask passphrase and unlock */
         ChangePass, /**< Ask old passphrase + new passphrase twice */
+        UnlockMigration, // Ask passphrase for unlocking during migration
```
Comment (hebasto) [inline src/qt/askpassphrasedialog.h:29]: Mind following the Doxygen comment formatting above?

File: src/interfaces/wallet.h [hunk 3]
```diff
@@ -46,6 +46,7 @@ enum isminetype : unsigned int;
 struct CRecipient;
 struct WalletContext;
 using isminefilter = std::underlying_type<isminetype>::type;
+enum class DatabaseFormat;
```
Comment (hebasto) [inline src/interfaces/wallet.h:49]: Why is this line needed?

**Comment by hebasto** [comment]
> This PR changes that menu item into a menu list like Open Wallet and lets users migrate any legacy wallet in their wallet directory regardless of the wallets loaded.

Tested d45ac03a89cc500cd461f878f092cf8ec99e7760. The "Migrate Wallet" menu item is still disabled when no wallet is loaded.

**Comment by hebasto** [comment]
Please rebase.

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (pablomartin4btc) [reply src/qt/bitcoingui.cpp:468]: Since #828 has been merged, please keep in mind this update...
```suggestion
                // Single & are shown when && is in the string. So replace & with &&.
                name.replace(QChar('&'), QString("&&"));
```

**Review by pablomartin4btc — COMMENTED** [review]
tACK d45ac03a89cc500cd461f878f092cf8ec99e7760

<details>
<summary>Created a couple of legacy wallets (encrypted and not) to begin with.</summary>

 <ins>**edit**</ins>: running ./configure with: `--with-bdb`.

Using `bitcoind` (`./src/bitcoind -regtest -deprecatedrpc=create_bdb
`) and `bitcoin-cli` (`./src/bitcoin-cli -regtest createwallet legacyWallet false false "" false false
`) or directly on the `rpc-console` (`createwallet legacyWalletEnc false false "passphrase" false false`) in `bitcoin-qt` (`./src/qt/bitcoin-qt -regtest -deprecatedrpc=create_bdb`).

```
{
  "name": "legacyWalletEnc",
  "warnings": [
    "Wallet created successfully. The legacy wallet type is being deprecated and support for creating and opening legacy wallets will be removed in the future."
  ]
}
```

```
{
  "name": "legacyWallet",
  "warnings": [
    "Empty string given as passphrase, wallet will not be encrypted.",
    "Wallet created successfully. The legacy wallet type is being deprecated and support for creating and opening legacy wallets will be removed in the future."
  ]
}
```

</details>

<details>
<summary>The menu item "Migrate Wallet" is shown when there are already wallets loaded.</summary>

image

</details>

<details>
<summary>But "Migrate Wallet" is disabled when there are no wallets loaded as @hebasto mentioned <a href="https://github.com/bitcoin-core/gui/pull/824#issuecomment-2227310588">above</a>.</summary>

image

</details>

- I think that should be solved by removing the migration action from `BitcoinGUI::setWalletActionsEnabled` (the migration action has the same behaviour as the restore action which is not there).

<details>
<summary>"Migrate Wallet" is shown when there are no wallets available at all.</summary>

image

</details>


<details>
<summary>Tested migration of encrypted wallet with and without providing the passphrase, working both cases.</summary>

image

image

</details>

<details>
<summary>Every migration failure on a particular wallet still creates a backup file. Do we want this?</summary>

image
</details>

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (achow101) [reply src/qt/bitcoingui.cpp:467]: Fixed during rebase.

File: src/qt/askpassphrasedialog.h (hunk 2 above)
Comment (achow101) [reply src/qt/askpassphrasedialog.h]: Done

File: src/interfaces/wallet.h (hunk 3 above)
Comment (achow101) [reply src/interfaces/wallet.h]: Removed

**Comment by achow101** [comment]
> Tested [d45ac03](https://github.com/bitcoin-core/gui/commit/d45ac03a89cc500cd461f878f092cf8ec99e7760). The "Migrate Wallet" menu item is still disabled when no wallet is loaded.

Should be fixed now.

**Comment by pablomartin4btc** [comment]
> > "Migrate Wallet" is shown when there are no wallets available at all.
> 
> I think that's ok.

Yeah, it was just an observation (mainly to compare when the menu is disabled), thanks!

**Review by pablomartin4btc — APPROVED** [review]
> > Tested [d45ac03](https://github.com/bitcoin-core/gui/commit/d45ac03a89cc500cd461f878f092cf8ec99e7760). The "Migrate Wallet" menu item is still disabled when no wallet is loaded.
> 
> Should be fixed now.
> 

re-ACK bcf47b9f34e4f21e5ac75521ddd4f25ff3812412

File: src/qt/walletcontroller.cpp [hunk 4]
```diff
@@ -436,47 +445,50 @@ void RestoreWalletActivity::finish()
...
+        AskPassphraseDialog dlg(AskPassphraseDialog::UnlockMigration, m_parent_widget, &passphrase);
+        if (dlg.exec() == QDialog::Rejected) return;
     }
 
     // GUI needs to remove the wallet so that it can actually be unloaded by migration
-    const std::string name = wallet_model->wallet().getWalletName();
-    m_wallet_controller->removeAndDeleteWallet(wallet_model);
+    if (WalletModel* wallet = m_wallet_controller->getWallet(name)) {
+        m_wallet_controller->removeAndDeleteWallet(wallet);
+    };
```
Comment (furszy) [inline src/qt/walletcontroller.cpp:478]: We don't actually need to do this. Could cherry-pick https://github.com/furszy/bitcoin-core/commit/f5c0c58977a8989fe17db361d179281d051b0806 and remove the introduced `getWallet`.

**Review by furszy — COMMENTED** [review]
Few comments. Will continue reviewing.

File: src/qt/bitcoingui.cpp [hunk 5]
```diff
@@ -455,10 +456,33 @@ void BitcoinGUI::createActions()
...
+                if (format != "bdb") {
+                    // This wallet is already migrated
+                    action->setEnabled(false);
+                    continue;
+                }
+
+                connect(action, &QAction::triggered, [this, wallet_name] {
+                    auto activity = new MigrateWalletActivity(m_wallet_controller, this);
+                    connect(activity, &MigrateWalletActivity::migrated, this, &BitcoinGUI::setCurrentWallet);
+                    activity->migrate(wallet_name);
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/qt/bitcoingui.cpp:479]: What compiler? I'm not seeing this error, and it seems neither did CI.

File: src/qt/bitcoingui.cpp [hunk 6]
```diff
@@ -455,10 +456,33 @@ void BitcoinGUI::createActions()
...
+                // the string, but only the first occurrence. So replace only
+                // the first & with &&.
+                name.replace(name.indexOf(QChar('&')), 1, QString("&&"));
+                QAction* action = m_migrate_wallet_menu->addAction(name);
+
+                if (format != "bdb") {
+                    // This wallet is already migrated
+                    action->setEnabled(false);
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/qt/bitcoingui.cpp]: Done

File: src/qt/walletcontroller.cpp (hunk 4 above)
Comment (achow101) [reply src/qt/walletcontroller.cpp]: Cherry picked.

File: src/qt/bitcoingui.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (furszy) [reply src/qt/bitcoingui.cpp:478]: > What compiler? I'm not seeing this error, and it seems neither did CI.

clang 14.0.3 (clang-1403.0.22.14.1).

**Comment by achow101** [comment]
> Could replace it directly for [furszy/bitcoin-core@e8d147e](https://github.com/furszy/bitcoin-core/commit/e8d147eeaacac7d434aad8dbbc5022aadf7d112a). It detects the wallet encryption status automatically regardless of its loaded/unloaded state.

Cherry picked.

Also fixed test failure.

**Review by furszy — COMMENTED** [review]
ACK 0240454120a88915396b0eb68d9b5bee73e52fed

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (hebasto) [reply src/qt/bitcoingui.cpp:472]: > Fixed during rebase.

It still needs to be fixed.

**Review by hebasto — COMMENTED** [review]
Almost ACK fd087d4667b94051d00617189dcecc4f88d849ce.

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (achow101) [reply src/qt/bitcoingui.cpp]: Ah, missed one, fixed.

**Review by hebasto — APPROVED** [review]
ACK 8f2522d242961ceb9e79672aa43e856863a1a6dd.

**Comment by hebasto** [comment]
cc @furszy

**Review by furszy — COMMENTED** [review]
ACK 8f2522d

File: src/qt/bitcoingui.cpp [hunk 7]
```diff
@@ -455,10 +456,31 @@ void BitcoinGUI::createActions()
...
+                QAction* action = m_migrate_wallet_menu->addAction(name);
+
+                connect(action, &QAction::triggered, [this, wallet_name] {
+                    auto activity = new MigrateWalletActivity(m_wallet_controller, this);
+                    connect(activity, &MigrateWalletActivity::migrated, this, &BitcoinGUI::setCurrentWallet);
+                    activity->migrate(wallet_name);
+                });
+            }
+            if (m_migrate_wallet_menu->isEmpty()) {
+                QAction* action = m_migrate_wallet_menu->addAction(tr("No wallets available"));
```
Comment (MarnixCroes) [inline src/qt/bitcoingui.cpp:481]: nit: maybe a more descriptive text would be good?
something like `No wallets that need to be migrated`

for people who don't know about the migration and why/what it is.
seeing _No wallets available_ while they have wallet(s) might seem wrong/confusing.

sry for commenting post-merge

**Review by pablomartin4btc — COMMENTED** [review]
post-merge tACK 8f2522d242961ceb9e79672aa43e856863a1a6dd

Checked changes proposed by @furszy.

<details>
<summary>Perhaps the PR's description needs to be updated (or a note added) since the previous extra button to enter the passphrase was removed (<code>isEncrypted</code> is in place). Nice solution, less prone to errors too.</summary>

image

</details>
